File: sdk/python/src/tinyhumans/http.py
from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any
from urllib.error import HTTPError
from urllib.parse import urlencode, urljoin
from urllib.request import Request, urlopen


Json = dict[str, Any] | list[Any] | str | int | float | bool | None


class TinyHumansError(RuntimeError):
    def __init__(self, status: int, body: Any, message: str | None = None) -> None:
        self.status = status
        self.body = body
        super().__init__(message or f"TinyHumans request failed with HTTP {status}")


@dataclass(slots=True)
class HttpClient:
    base_url: str
    token: str | None = None
    api_key: str | None = None
    admin_service_token: str | None = None
    headers: dict[str, str] | None = None
    unwrap_envelope: bool = True
    timeout: float = 30.0

    def request(
        self,
        method: str,
        path: str,
        *,
        query: dict[str, Any] | None = None,
        body: Any = None,
        headers: dict[str, str] | None = None,
        unwrap_envelope: bool | None = None,
    ) -> Json:
        url = self._url(path, query)
        request_headers = self._headers(headers)
        data: bytes | None = None
        if body is not None:
            request_headers.setdefault("content-type", "application/json")
            data = body.encode("utf-8") if isinstance(body, str) else json.dumps(body).encode("utf-8")

        request = Request(url, data=data, headers=request_headers, method=method.upper())
        try:
            with urlopen(request, timeout=self.timeout) as response:
                parsed = _read_body(response.read(), response.headers.get_content_type())
                if response.status == 204:
                    return None
                return self._unwrap(parsed, unwrap_envelope)
        except HTTPError as error:
            parsed = _read_body(error.read(), error.headers.get_content_type())
            message = parsed.get("error") if isinstance(parsed, dict) else None
            raise TinyHumansError(error.code, parsed, str(message) if message else None) from error

    def get(self, path: str, **kwargs: Any) -> Json:
        return self.request("GET", path, **kwargs)

    def post(self, path: str, body: Any = None, **kwargs: Any) -> Json:
        return self.request("POST", path, body=body, **kwargs)

    def put(self, path: str, body: Any = None, **kwargs: Any) -> Json:
        return self.request("PUT", path, body=body, **kwargs)

    def patch(self, path: str, body: Any = None, **kwargs: Any) -> Json:
        return self.request("PATCH", path, body=body, **kwargs)

    def delete(self, path: str, **kwargs: Any) -> Json:
        return self.request("DELETE", path, **kwargs)

    def _url(self, path: str, query: dict[str, Any] | None) -> str:
        base = self.base_url.rstrip("/") + "/"
        relative = path.lstrip("/")
        url = urljoin(base, relative)
        if query:
            pairs: list[tuple[str, str]] = []
            for key, value in query.items():
                values = value if isinstance(value, list) else [value]
                pairs.extend((key, str(item)) for item in values if item is not None)
            if pairs:
                url = f"{url}?{urlencode(pairs)}"
        return url

    def _headers(self, headers: dict[str, str] | None) -> dict[str, str]:
        merged = {
            "accept": "application/json",
            "x-sdk-client": "tinyhumans-python",
            **(self.headers or {}),
            **(headers or {}),
        }
        if self.token:
            merged["authorization"] = f"Bearer {self.token}"
        if self.api_key:
            merged["x-api-key"] = self.api_key
        if self.admin_service_token:
            merged["x-admin-service-token"] = self.admin_service_token
        return merged

    def _unwrap(self, body: Json, unwrap_envelope: bool | None) -> Json:
        should_unwrap = self.unwrap_envelope if unwrap_envelope is None else unwrap_envelope
        if should_unwrap and isinstance(body, dict) and body.get("success") is True and "data" in body:
            return body["data"]
        return body


def _read_body(data: bytes, content_type: str | None) -> Json:
    if not data:
        return None
    text = data.decode("utf-8")
    if content_type == "application/json" or text[:1] in ("{", "["):
        try:
            return json.loads(text)
        except json.JSONDecodeError:
            pass
    return text

File: sdk/typescript/src/http.ts
export type QueryValue = string | number | boolean | null | undefined;
export type Query = Record<string, QueryValue | QueryValue[]>;

export interface TinyHumansClientOptions {
  baseUrl: string;
  token?: string;
  apiKey?: string;
  headers?: Record<string, string>;
  fetch?: typeof globalThis.fetch;
  unwrapEnvelope?: boolean;
}

export interface RequestOptions {
  query?: Query;
  body?: unknown;
  headers?: Record<string, string>;
  responseType?: "json" | "text" | "raw";
  unwrapEnvelope?: boolean;
}

export class TinyHumansError extends Error {
  constructor(
    public readonly status: number,
    public readonly body: unknown,
    public readonly headers: Headers,
    message?: string,
  ) {
    super(message ?? `TinyHumans request failed with HTTP ${status}`);
    this.name = "TinyHumansError";
  }

  toJSON(): Record<string, unknown> {
    return {
      name: this.name,
      message: this.message,
      status: this.status,
      body: this.body,
    };
  }
}

export class HttpClient {
  private readonly baseUrl: string;
  private readonly fetchImpl: typeof globalThis.fetch;
  private readonly unwrapEnvelope: boolean;

  constructor(private readonly options: TinyHumansClientOptions) {
    this.baseUrl = options.baseUrl.replace(/\/+$/, "");
    this.fetchImpl = options.fetch ?? globalThis.fetch;
    this.unwrapEnvelope = options.unwrapEnvelope ?? true;

    if (!this.fetchImpl) {
      throw new Error("TinyHumansClient requires a fetch implementation");
    }
  }

  async request<T = unknown>(
    method: string,
    path: string,
    options: RequestOptions = {},
  ): Promise<T> {
    const url = this.buildUrl(path, options.query);
    const headers = this.buildHeaders(options);
    const init: RequestInit = { method, headers };

    if (options.body !== undefined) {
      headers.set("content-type", headers.get("content-type") ?? "application/json");
      init.body =
        typeof options.body === "string" ||
        options.body instanceof FormData ||
        options.body instanceof Blob
          ? (options.body as BodyInit)
          : JSON.stringify(options.body);
    }

    const response = await this.fetchImpl(url, init);
    if (options.responseType === "raw") {
      if (!response.ok) {
        throw new TinyHumansError(response.status, await safeBody(response), response.headers);
      }
      return response as T;
    }

    const body =
      options.responseType === "text" ? await response.text() : await safeBody(response);

    if (!response.ok) {
      const message =
        typeof body === "object" && body && "error" in body
          ? String((body as { error?: unknown }).error)
          : undefined;
      throw new TinyHumansError(response.status, body, response.headers, message);
    }

    if (response.status === 204) return undefined as T;

    const shouldUnwrap = options.unwrapEnvelope ?? this.unwrapEnvelope;
    if (
      shouldUnwrap &&
      body &&
      typeof body === "object" &&
      "success" in body &&
      (body as { success?: unknown }).success === true &&
      "data" in body
    ) {
      return (body as { data: T }).data;
    }

    return body as T;
  }

  get<T = unknown>(path: string, options?: RequestOptions): Promise<T> {
    return this.request<T>("GET", path, options);
  }

  post<T = unknown>(path: string, body?: unknown, options: RequestOptions = {}): Promise<T> {
    return this.request<T>("POST", path, { ...options, body });
  }

  put<T = unknown>(path: string, body?: unknown, options: RequestOptions = {}): Promise<T> {
    return this.request<T>("PUT", path, { ...options, body });
  }

  patch<T = unknown>(path: string, body?: unknown, options: RequestOptions = {}): Promise<T> {
    return this.request<T>("PATCH", path, { ...options, body });
  }

  delete<T = unknown>(path: string, options?: RequestOptions): Promise<T> {
    return this.request<T>("DELETE", path, options);
  }

  private buildUrl(path: string, query?: Query): URL {
    const normalizedPath = path.startsWith("/") ? path : `/${path}`;
    const url = new URL(`${this.baseUrl}${normalizedPath}`);
    for (const [key, value] of Object.entries(query ?? {})) {
      const values = Array.isArray(value) ? value : [value];
      for (const item of values) {
        if (item !== undefined && item !== null) url.searchParams.append(key, String(item));
      }
    }
    return url;
  }

  private buildHeaders(options: RequestOptions): Headers {
    const headers = new Headers({
      "accept": "application/json",
      "x-sdk-client": "@tinyhumansai/sdk",
      ...this.options.headers,
      ...options.headers,
    });
    if (this.options.token) headers.set("authorization", `Bearer ${this.options.token}`);
    if (this.options.apiKey) headers.set("x-api-key", this.options.apiKey);
    return headers;
  }
}

async function safeBody(response: Response): Promise<unknown> {
  const text = await response.text();
  if (!text) return undefined;
  try {
    return JSON.parse(text);
  } catch {
    return text;
  }
}
